#include "mainwindow.h"
#include "ui_mainwindow.h"
#include "platform/CameraHelper.h"
#include <QTcpSocket>
#include <QMessageBox>       // ç”¨äºå±•ç¤ºä¿¡æ¯æ¡†
#include <QDateTime>         // ç”¨äºç”Ÿæˆå”¯ä¸€çš„æ–‡ä»¶å
#include <QDebug>
#include <QNetworkProxy>

MainWindow::MainWindow(QWidget *parent)
    : QMainWindow(parent)
    , ui(new Ui::MainWindow)
{
    ui->setupUi(this);

    // åˆå§‹åŒ–Socket
    m_socket = new QTcpSocket(this);
    // [Engineering Fix] 2025-12-30
    // é—®é¢˜ï¼šå¼€å¯ç³»ç»Ÿä»£ç†(VPN)æ—¶ï¼ŒQTcpSocket ä¼šå°è¯•é€šè¿‡ä»£ç†æœåŠ¡å™¨è¿æ¥å†…ç½‘ IPï¼Œå¯¼è‡´è¿æ¥è¶…æ—¶ã€‚
    // ç°è±¡ï¼šèƒ½ Ping é€š (ICMP åè®®ä¸èµ°ä»£ç†)ï¼Œä½† TCP è¿æ¥å¤±è´¥ã€‚
    // è§£å†³ï¼šå¼ºåˆ¶è®¾ç½® NoProxyï¼Œç¡®ä¿ Socket ç›´è¿ç‰©ç†ç½‘å¡ã€‚
    m_socket->setProxy(QNetworkProxy::NoProxy);

    // TCPé€šä¿¡ä¿¡å·ä¸æ§½è¿æ¥
    connect(m_socket, &QTcpSocket::connected, this, &MainWindow::onSocketConnected);
    connect(m_socket, &QTcpSocket::disconnected, this, &MainWindow::onSocketDisconnected);
    connect(m_socket, &QTcpSocket::errorOccurred, this, &MainWindow::onSocketError);
    // æœºæ¢°è‡‚æ§åˆ¶ä¿¡å·ä¸æ§½è¿æ¥
    connect(ui->btn_X_Plus, &QPushButton::pressed, this, [=](){ onJogBtnPressed(0, 1); });
    connect(ui->btn_X_Plus, &QPushButton::released, this, &MainWindow::onJogBtnReleased);
    connect(ui->btn_X_Minus, &QPushButton::pressed, this, [=](){ onJogBtnPressed(0, -1); });
    connect(ui->btn_X_Minus, &QPushButton::released, this, &MainWindow::onJogBtnReleased);
    connect(ui->btn_Y_Plus, &QPushButton::pressed, this, [=](){ onJogBtnPressed(1, 1); });
    connect(ui->btn_Y_Plus, &QPushButton::released, this, &MainWindow::onJogBtnReleased);
    connect(ui->btn_Y_Minus, &QPushButton::pressed, this, [=](){ onJogBtnPressed(1, -1); });
    connect(ui->btn_Y_Minus, &QPushButton::released, this, &MainWindow::onJogBtnReleased);
    connect(ui->btn_Z_Plus, &QPushButton::pressed, this, [=](){ onJogBtnPressed(2, 1); });
    connect(ui->btn_Z_Plus, &QPushButton::released, this, &MainWindow::onJogBtnReleased);
    connect(ui->btn_Z_Minus, &QPushButton::pressed, this, [=](){ onJogBtnPressed(2, -1); });
    connect(ui->btn_Z_Minus, &QPushButton::released, this, &MainWindow::onJogBtnReleased);

    // åˆå§‹åŒ–ç•Œé¢çŠ¶æ€
    ui->lbl_Status->setText("æœªè¿æ¥");
    ui->lbl_Status->setStyleSheet("color:red;");

    ui->lbl_Cam1->setStyleSheet("QLabel { background-color: black; }");
    ui->lbl_Cam2->setStyleSheet("QLabel { background-color: black; }");
    ui->lbl_Cam3->setStyleSheet("QLabel { background-color: black; }");
    ui->lbl_Cam4->setStyleSheet("QLabel { background-color: black; }");

    // åˆå§‹åŒ–ç›¸æœº
    int cameraCount = 4;

    for(int i = 0; i < cameraCount; i++){
        cv::VideoCapture cap = createCamera(i);
        if(cap.isOpened()){
            // æ‰“å°æœ€ç»ˆå®é™…è·å–åˆ°çš„åˆ†è¾¨ç‡ (ç”¨äºéªŒè¯)
            double actualW = cap.get(cv::CAP_PROP_FRAME_WIDTH);
            double actualH = cap.get(cv::CAP_PROP_FRAME_HEIGHT);
            qDebug() << "âœ… ç›¸æœº" << i << "åˆå§‹åŒ–æˆåŠŸ | åˆ†è¾¨ç‡:" << actualW << "x" << actualH;

            m_cams.push_back(cap);
        } else {
            m_cams.push_back(cv::VideoCapture());     // å³ä½¿æ‰“ä¸å¼€ï¼Œä¹Ÿè¦å‹å…¥ä¸€ä¸ªç©ºå¯¹è±¡å ä½ï¼Œé˜²æ­¢åé¢æ•°ç»„è¶Šç•Œ
        } 

        m_currentFrames.push_back(cv::Mat());
    }

    // å¯åŠ¨å®šæ—¶å™¨
    m_timer = new QTimer(this);
    connect(m_timer, &QTimer::timeout, this, &MainWindow::updateFrames);
    m_timer->start(33); // 33ms â‰ˆ 30 FPS
}

MainWindow::~MainWindow()
{
    // ç¨‹åºå…³é—­å‰é‡Šæ”¾ç›¸æœºèµ„æº
    for(auto &cap : m_cams){
        if(cap.isOpened()) cap.release();
    }

    delete ui;
}

// ç‚¹å‡»æŒ‰é’®æ§½å‡½æ•°
void MainWindow::on_btn_Connect_clicked()
{
    // å¦‚æœå½“å‰å·²ç»æ˜¯è¿æ¥çŠ¶æ€ï¼Œç‚¹å‡»æŒ‰é’®æ„å‘³ç€"æ–­å¼€è¿æ¥"
    if(m_socket->state() == QAbstractSocket::ConnectedState){
        ui->btn_Connect->setEnabled(false); // é˜²æ­¢é‡å¤ç‚¹å‡»

        // m_socket->disconnectFromHost();    // ç¤¼è²Œåˆ†æ‰‹ï¼Œç­‰å¾…æœåŠ¡å™¨æ–¹ç¡®è®¤æ‰ä¼šè¿›å…¥UnconnectedStateã€‚ç°åœ¨æ˜¯ClosingState

        // abort(): æ‹”ç½‘çº¿ï¼Œç«‹å³å…³é—­è¿æ¥ï¼Œé‡ç½® Socketï¼ˆå¿«ï¼Œç«‹å³è¿›å…¥ UnconnectedStateï¼‰
        // abort()æ–­å¼€åï¼Œç¨‹åºä¼šç«‹å³æ‰§è¡ŒonSocketDisconnected()ï¼Œæ‰€ä»¥è¦å…ˆæ”¹å˜UIæ–‡æœ¬ï¼Œåœ¨æ‰§è¡Œæ–­å¼€æ“ä½œ
        ui->lbl_Status->setText("æ­£åœ¨æ–­å¼€...");
        m_socket->abort();
        return;
    }

    QString ip = ui->le_IPAdress->text();
    quint16 port = 30003;

    // è¾“å…¥æ ¡éªŒ
    if (ip.isEmpty()){
        QMessageBox::warning(this, "æç¤º", "IPåœ°å€ä¸èƒ½ä¸ºç©º");
        return;
    }

    ui->lbl_Status->setText("æ­£åœ¨è¿æ¥...");
    ui->btn_Connect->setEnabled(false);     // è¿æ¥è¿‡ç¨‹ä¸­ç¦ç”¨æŒ‰é’®ï¼Œé˜²æ­¢ç‹‚ç‚¹

    // å‘èµ·è¿æ¥
    m_socket->connectToHost(ip, port);
}

// è¿æ¥æˆåŠŸæ§½å‡½æ•°
void MainWindow::onSocketConnected()
{
    ui->lbl_Status->setText("è¿æ¥æˆåŠŸ");
    ui->lbl_Status->setStyleSheet("color: green; font-weight: bold;");

    ui->btn_Connect->setText("æ–­å¼€è¿æ¥");
    ui->btn_Connect->setEnabled(true);      // æŒ‰é’®å¯ä»¥æŒ‰ä¸‹
}

// æ–­å¼€è¿æ¥æ§½å‡½æ•°
void MainWindow::onSocketDisconnected()
{
    ui->lbl_Status->setText("å·²æ–­å¼€");
    ui->lbl_Status->setStyleSheet("color: red;");

    ui->btn_Connect->setText("è¿æ¥");
    ui->btn_Connect->setEnabled(true);
}

// è¿æ¥é”™è¯¯æ§½å‡½æ•°
void MainWindow::onSocketError(QAbstractSocket::SocketError socketError)
{
    // è·å–å¯è¯»çš„é”™è¯¯ä¿¡æ¯
    QString errorMsg = m_socket->errorString();

    ui->lbl_Status->setText("é”™è¯¯: " + errorMsg);
    ui->lbl_Status->setStyleSheet("color: darkred;");

    ui->btn_Connect->setText("è¿æ¥æœºæ¢°è‡‚");
    ui->btn_Connect->setEnabled(true);

    // åªæœ‰éæ–­å¼€å¼•èµ·çš„é”™è¯¯æ‰å¼¹çª—ï¼Œä½“éªŒæ›´å¥½
    if (socketError != QAbstractSocket::RemoteHostClosedError) {
        QMessageBox::critical(this, "è¿æ¥é”™è¯¯", "æ— æ³•è¿æ¥åˆ°æœºæ¢°è‡‚:\n" + errorMsg);
    }
}

// å®šæ—¶åˆ·æ–°é€»è¾‘
void MainWindow::updateFrames()
{
    // å°† UI ä¸Šçš„æ ‡ç­¾æ”¾å…¥æ•°ç»„ï¼Œæ–¹ä¾¿å¾ªç¯æ“ä½œ
    QLabel* displayLabels[] = {ui->lbl_Cam1, ui->lbl_Cam2, ui->lbl_Cam3, ui->lbl_Cam4};

    // éå†æ‰€æœ‰å·²ç®¡ç†çš„ç›¸æœº
    for(size_t i = 0; i < m_cams.size(); i++) {

        if(m_cams[i].isOpened()) {
            cv::Mat frame;
            m_cams[i] >> frame; // æŠ“å–ä¸€å¸§

            if(!frame.empty()) {
                // 1. å­˜å…¥ç¼“å­˜ï¼ˆå¿…é¡»å­˜åŸå§‹ BGR æ•°æ®ï¼Œç”¨äºä¿å­˜å›¾ç‰‡ï¼‰
                m_currentFrames[i] = frame.clone();

                // 2. è½¬æ¢å¹¶æ˜¾ç¤º
                QImage qimg = matToQImage(frame);
                displayLabels[i]->setPixmap(QPixmap::fromImage(qimg));
            }
        }
    }
}

// è¾…åŠ©å‡½æ•°ï¼šMat (OpenCV) -> QImage (Qt)
QImage MainWindow::matToQImage(const cv::Mat &mat)
{
    // 1. é¢œè‰²è½¬æ¢ BGR -> RGB
    cv::Mat rgb;
    cv::cvtColor(mat, rgb, cv::COLOR_BGR2RGB);

    // 2. æ„é€  QImage
    QImage img((const uchar*)rgb.data, rgb.cols, rgb.rows, rgb.step, QImage::Format_RGB888);

    // 3. é‡è¦ï¼šæ·±æ‹·è´
    // å¿…é¡» copy()ï¼Œå› ä¸º rgb æ˜¯å±€éƒ¨å˜é‡ï¼Œå‡½æ•°ç»“æŸåå†…å­˜ä¼šè¢«é‡Šæ”¾
    return img.copy();
}

// æˆªå›¾ä¿å­˜æŒ‰é’®
void MainWindow::on_btn_Capture_clicked()
{
    // ä½¿ç”¨å½“å‰æ—¶é—´ç”Ÿæˆæ–‡ä»¶åï¼Œç²¾ç¡®åˆ°ç§’
    QString timestamp = QDateTime::currentDateTime().toString("yyyyMMdd_HHmmss");
    bool savedAny = false;

    for(size_t i = 0; i < m_currentFrames.size(); i++) {
        if(!m_currentFrames[i].empty()) {
            // æ–‡ä»¶åç¤ºä¾‹: Cam1_20251217_203000.jpg
            QString filename = QString("Cam%1_%2.jpg").arg(i+1).arg(timestamp);

            // ä½¿ç”¨ OpenCV ä¿å­˜å›¾ç‰‡ (è´¨é‡å¥½ï¼Œä¸”å…¼å®¹æ€§å¼º)
            // æ³¨æ„ï¼šimwrite éœ€è¦ std::string
            cv::imwrite(filename.toStdString(), m_currentFrames[i]);

            qDebug() << "å·²ä¿å­˜:" << filename;
            savedAny = true;
        }
    }

    if(savedAny) {
        // çŠ¶æ€æ æç¤ºä¸€ä¸‹å³å¯ï¼Œä¸å¼¹çª—æ‰“æ‰°æ“ä½œ
        ui->lbl_Status->setText("æˆªå›¾å·²ä¿å­˜è‡³è¿è¡Œç›®å½•");
    } else {
        QMessageBox::warning(this, "è­¦å‘Š", "å½“å‰æ²¡æœ‰å›¾åƒæ•°æ®ï¼Œæ— æ³•ä¿å­˜ï¼");
    }
}

// æœºæ¢°è‡‚æ§åˆ¶
// 1. æŒ‡ä»¤å‘é€å‡½æ•°
void MainWindow::sendURScript(QString cmd)
{
    if(m_socket->state() != QAbstractSocket::ConnectedState) {
        qDebug() << "âš ï¸ æœªè¿æ¥æœºæ¢°è‡‚ï¼ŒæŒ‡ä»¤å‘é€å¤±è´¥";
        return;
    }

    // URScript å¿…é¡»ä»¥æ¢è¡Œç¬¦ '\n' ç»“å°¾ï¼Œå¦åˆ™æœºå™¨ä¸æ‰§è¡Œ
    cmd.append("\n");

    QByteArray data = cmd.toUtf8();
    qint64 bytesWritten = m_socket->write(data);

    if(bytesWritten == -1) {
        qDebug() << "âš ï¸ å†™å…¥å¤±è´¥";
        return;
    } else if(bytesWritten != data.size()) {
        qDebug() << "âš ï¸ å†™å…¥éƒ¨åˆ†æ•°æ®ï¼Œå®é™…å†™å…¥" << bytesWritten << "å­—èŠ‚";
        return;
    } else {
        m_socket->flush();  // ç¡®ä¿ç«‹å³å‘é€ç¼“å†²åŒºæ•°æ®
    }
}

// 2. å®ç°æŒ‰ä¸‹æŒ‰é’®ï¼ˆå¼€å§‹ç§»åŠ¨ï¼‰
// axis: 0=X, 1=Y, 2=Z
// direction: 1 or -1
void MainWindow::onJogBtnPressed(int axis, int direction)
{
    // æ„å»ºé€Ÿåº¦å‘é‡ [Vx, Vy, Vz, Rx, Ry, Rz]
    double speeds[6] = {0, 0, 0, 0, 0, 0};
    speeds[axis] = direction * MOVE_VEL;

    // æ‹¼æ¥ URScript å­—ç¬¦ä¸²: speedl([x,y,z,rx,ry,rz], a, t)
    // t è®¾ç½®ä¸º 100ç§’ï¼Œæ„å‘³ç€â€œä¸€ç›´åŠ¨ä¸‹å»â€ï¼Œç›´åˆ°å‘ stopl
    QString script = QString("speedl([%1, %2, %3, 0, 0, 0], %4, 100)")
                         .arg(speeds[0]).arg(speeds[1]).arg(speeds[2])
                         .arg(MOVE_ACC);

    qDebug() << "ğŸ“¤ å‘é€æŒ‡ä»¤:" << script;
    sendURScript(script);
}

// 3. å®ç°æ¾å¼€æŒ‰é’®ï¼ˆç«‹å³åœæ­¢ï¼‰
void MainWindow::onJogBtnReleased()
{
    // stopl(a): çº¿æ€§åœæ­¢
    QString script = QString("stopl(%1)").arg(MOVE_ACC);

    qDebug() << "ğŸ›‘ å‘é€åœæ­¢";
    sendURScript(script);
}




